qermit/taskgraph/mitres.py

```diff
@@ -302,3 +302,141 @@ def run(self, circuit_shots: List[CircuitShots]) -> List[BackendResult]:  # type
+        # this is the case.
+        for i, circ_shots in enumerate(circuit_wires):
+
+            shots_to_run = circ_shots.Shots
+
+            while shots_to_run > 0:
+
+                split_circ_shots = CircuitShots(
+                    circ_shots.Circuit, min(max_shots, shots_to_run)
+                )
+                # append a job with a reduced number of shots to the list.
+                split_circuit_wires.append(split_circ_shots)
+                # Append the original circuit index to the index list.
+                split_index.append(i)
+                shots_to_run -= max_shots
+
+        return (split_circuit_wires, split_index)
+
+    return MitTask(_label="SplitShots", _n_in_wires=1, _n_out_wires=2, _method=task)
+
+
+def group_shots_task_gen() -> MitTask:
+    """Returns task which groups together jobs that correspond to the same
+    circuit. Original larger job could have been split by task returned
+    by split_shots_task_gen.
+
+    :return: Task merging jobs which correspond to the same circuit.
+    :rtype: MitTask
+    """
+
+    def task(
+        obj, results_wires: List[BackendResult], split_index: List[int]
+    ) -> Tuple[List[BackendResult]]:
+        """Task which merges jobs which correspond to the same circuit.
+
+        :param results_wires: A list of results, some of which correspond to
+        the same circuit.
+        :type results_wires: List[BackendResult]
+        :param split_index: A list of indexes identifying which results
+        correspond to which circuit. Those results with index i correspond to
+        the circuit i.
+        :type split_index: List[int]
+        :return: A new list of results, where results which correspond to the
+        same circuit have been grouped together.
+        :rtype: Tuple[List[BackendResult]]
+        """
+
+        # A new list of results where some results have been merged.
+        merged_results = []
+
+        # A list of lists. Each internal list contains results that correspond
+        # to the circuit of the same index.
+        grouped_results = [
+            [result for result, i in zip(results_wires, split_index) if i == j]
+            for j in range(split_index[-1] + 1)
+        ]
+
+        # For each circuit, combine the results.
+        for fixed_circ in grouped_results:
+            # initialise result with the first that corresponds to this circuit.
+            circ_all_result_dict = fixed_circ[0].to_dict()
+            # For the remaining results, append the shots.
+            for circ_result in fixed_circ[1:]:
+                circ_all_result_dict["shots"]["array"].extend(
+                    circ_result.to_dict()["shots"]["array"]
+                )
+            merged_results.append(BackendResult().from_dict(circ_all_result_dict))
+
+        return (merged_results,)
+
+    return MitTask(_label="MergeShots", _n_in_wires=2, _n_out_wires=1, _method=task)
+
+
+def gen_shot_split_mitres(backend: Backend, max_shots: int) -> MitRes:
```

Comment:
it would be nice to be able to define this type of mitres and also have compilation as an option - dya think thats doable?
Also for consistency please change to `gen_shot_split_MitRes`